src/type/validate.js

```diff
@@ -314,94 +317,110 @@ function validateFields(
   }
 }
 
-function validateObjectInterfaces(
+function validateInterfaces(
   context: SchemaValidationContext,
-  object: GraphQLObjectType,
+  implementing: GraphQLObjectType | GraphQLInterfaceType,
 ): void {
   const implementedTypeNames = Object.create(null);
-  for (const iface of object.getInterfaces()) {
-    if (!isInterfaceType(iface)) {
+  for (const implemented of implementing.getInterfaces()) {
+    if (!isInterfaceType(implemented)) {
       context.reportError(
-        `Type ${inspect(object)} must only implement Interface types, ` +
-          `it cannot implement ${inspect(iface)}.`,
-        getAllImplementsInterfaceNodes(object, iface),
+        `Type ${inspect(implementing)} must only implement Interface types, ` +
+          `it cannot implement ${inspect(implemented)}.`,
+        getAllImplementsInterfaceNodes(implementing, implemented),
       );
       continue;
     }
 
-    if (implementedTypeNames[iface.name]) {
+    if (implementedTypeNames[implemented.name]) {
       context.reportError(
-        `Type ${object.name} can only implement ${iface.name} once.`,
-        getAllImplementsInterfaceNodes(object, iface),
+        `Type ${implementing.name} can only implement ${implemented.name} once.`,
+        getAllImplementsInterfaceNodes(implementing, implemented),
       );
       continue;
     }
-    implementedTypeNames[iface.name] = true;
-    validateObjectImplementsInterface(context, object, iface);
+    implementedTypeNames[implemented.name] = true;
+    validateImplementsInterface(context, implementing, implemented);
   }
 }
 
-function validateObjectImplementsInterface(
+function validateImplementsInterface(
   context: SchemaValidationContext,
-  object: GraphQLObjectType,
-  iface: GraphQLInterfaceType,
+  implementing: GraphQLObjectType | GraphQLInterfaceType,
+  implemented: GraphQLInterfaceType,
 ): void {
-  const objectFieldMap = object.getFields();
-  const ifaceFieldMap = iface.getFields();
+  const implementingFieldMap = implementing.getFields();
+  const implementedFieldMap = implemented.getFields();
+
+  // Assert each ancestor interface is explicitly implemented.
+  validateImplementsAncestors(context, implementing, implemented);
```

Comment:
No need to make call stack deeper, please move it into `validateInterfaces` before `validateImplementsInterface` call.